include/sleip/dynamic_array.hpp

```diff
@@ -0,0 +1,320 @@
+  {
+    auto const count = static_cast<size_type>(std::distance(first, last));
+
+    data_ = std::allocator_traits<Allocator>::allocate(alloc_, count);
+
+#ifdef BOOST_NO_EXCEPTIONS
+    boost::alloc_construct_n(alloc_, data_, count, first);
+    size_ = count;
+#else
+    try {
+      boost::alloc_construct_n(alloc_, data_, count, first);
+      size_ = count;
+    }
+    catch (std::exception const& ex) {
+      std::allocator_traits<Allocator>::deallocate(alloc_, data_, count);
+      boost::throw_exception(ex);
+    }
+    catch (...) {
+      std::allocator_traits<Allocator>::deallocate(alloc_, data_, count);
+      throw;
+    }
+#endif
+  }
+
+  dynamic_array(dynamic_array const& other)
+    : alloc_(std::allocator_traits<allocator_type>::select_on_container_copy_construction(
+        other.get_allocator()))
+  {
+    data_ = std::allocator_traits<Allocator>::allocate(alloc_, other.size());
+
+#ifdef BOOST_NO_EXCEPTIONS
+    boost::alloc_construct_n(alloc_, data_, other.size(), other.begin());
+    size_ = other.size();
+#else
+    try {
+      boost::alloc_construct_n(alloc_, data_, other.size(), other.begin());
+      size_ = other.size();
+    }
+    catch (std::exception const& ex) {
+      std::allocator_traits<Allocator>::deallocate(alloc_, data_, other.size());
+      boost::throw_exception(ex);
+    }
+    catch (...) {
+      std::allocator_traits<Allocator>::deallocate(alloc_, data_, other.size());
+      throw;
+    }
+#endif
+  }
+
+  dynamic_array(dynamic_array const& other, Allocator const& alloc)
+    : alloc_(alloc)
+  {
+    data_ = std::allocator_traits<Allocator>::allocate(alloc_, other.size());
+
+#ifdef BOOST_NO_EXCEPTIONS
+    boost::alloc_construct_n(alloc_, data_, other.size(), other.begin());
+    size_ = other.size();
+#else
+    try {
+      boost::alloc_construct_n(alloc_, data_, other.size(), other.begin());
+      size_ = other.size();
+    }
+    catch (std::exception const& ex) {
+      std::allocator_traits<Allocator>::deallocate(alloc_, data_, other.size());
+      boost::throw_exception(ex);
+    }
+    catch (...) {
+      std::allocator_traits<Allocator>::deallocate(alloc_, data_, other.size());
+      throw;
+    }
+#endif
+  }
+
+  dynamic_array(dynamic_array&& other) noexcept
+    : data_(other.data_)
+    , size_(other.size_)
+    , alloc_(std::move(other.alloc_))
+  {
+    other.data_  = nullptr;
+    other.size_  = 0;
+    other.alloc_ = Allocator();
+  }
+
+  dynamic_array(dynamic_array&& other, Allocator const& alloc)
+    : alloc_(alloc)
+  {
+    if (alloc_ == other.alloc_) {
+      data_ = other.data_;
+      size_ = other.size_;
+
+      other.data_ = nullptr;
+      other.size_ = 0;
+      return;
+    }
+
+    data_ = std::allocator_traits<Allocator>::allocate(alloc_, other.size_);
+
+#ifdef BOOST_NO_EXCEPTIONS
+    for (size_type idx = 0; idx < other.size_; ++idx) {
+      std::allocator_traits<Allocator>::construct(alloc_, data_, std::move(other.data_[idx]));
+    }
```

Comment:
Why isn't this using `boost::alloc_construct_n`?

Reply:
Hmm, I wasn't sure how to handle this. Basically, if move wasn't noexcept, I wouldn't be able to hold up the strong exception guarantee without using a copy and the regular move iterator doesn't come with move_if_noexcept semantics.

So I rewrote the code to be more manual but work with move_if_noexcept.

Reply:
I've added a helper: `alloc_move_construct_n` which will `std::move_if_noexcept` `n` times.

This helper removed the last explicit try-catches in the source code!